File: v20_flyweight/java/DrawingApp2.java
package v20_flyweight.java;

import java.util.HashMap;
import java.util.Map;

interface Shape {
    void draw(int x, int y, int radius);
}

class CircleFlyweight implements Shape {
    private String color;

    public CircleFlyweight(String color) {
        this.color = color;
    }

    @Override
    public void draw(int x, int y, int radius) {
        System.out.printf("Drawing circle of color %s at position (%d, %d) with radius %d\n", color, x, y, radius);
    }
}

class ShapeFactory {
    private static final Map<String, CircleFlyweight> circleMap = new HashMap<>();

    public static CircleFlyweight getCircle(String color) {
        CircleFlyweight circle = circleMap.get(color);
        
        if (circle == null) {
            circle = new CircleFlyweight(color);
            circleMap.put(color, circle);
            System.out.println("Creating circle of color: " + color);
        }
        
        return circle;
    }
}

public class DrawingApp2 {
    public static void main(String[] args) {
        CircleFlyweight redCircle = ShapeFactory.getCircle("red");
        redCircle.draw(10, 10, 50);

        CircleFlyweight blueCircle = ShapeFactory.getCircle("blue");
        blueCircle.draw(20, 20, 60);

        CircleFlyweight anotherRedCircle = ShapeFactory.getCircle("red");
        anotherRedCircle.draw(30, 30, 40);
    }
}

File: v20_flyweight/python/drawing_app2.py
from typing import Dict
from abc import ABC, abstractmethod

class Shape(ABC):
    @abstractmethod
    def draw(self, x: int, y: int, radius: int):
        pass

class CircleFlyweight(Shape):
    def __init__(self, color: str):
        self.color = color

    def draw(self, x: int, y: int, radius: int):
        print(f'Drawing circle of color {self.color} at position ({x}, {y}) with radius {radius}')

class ShapeFactory:
    _circle_map: Dict[str, CircleFlyweight] = {}

    @staticmethod
    def get_circle(color: str) -> CircleFlyweight:
        circle = ShapeFactory._circle_map.get(color)

        if circle is None:
            circle = CircleFlyweight(color)
            ShapeFactory._circle_map[color] = circle
            print(f'Creating circle of color: {color}')

        return circle

class DrawingApp2:
    @staticmethod
    def main():
        red_circle = ShapeFactory.get_circle("red")
        red_circle.draw(10, 10, 50)

        blue_circle = ShapeFactory.get_circle("blue")
        blue_circle.draw(20, 20, 60)

        another_red_circle = ShapeFactory.get_circle("red")
        another_red_circle.draw(30, 30, 40)

if __name__ == "__main__":
    DrawingApp2.main()
